WooCommerce/Classes/ViewRelated/Products/Variations/Bulk Update/BulkUpdateOptionsModel.swift

```diff
@@ -0,0 +1,46 @@
+import Foundation
+import Yosemite
+
+/// Contains the logic for calculating if an array of `ProductVariation` has the same, different or missing values for a specific property.
+///
+final class BulkUpdateOptionsModel {
+    private(set) var productVariations: [ProductVariation]
+
+    init(productVariations: [ProductVariation]) {
+        self.productVariations = productVariations
+    }
+
+    /// Represents if a property in a collection of `ProductVariation`  has the same value or different values or is missing.
+    ///
+    enum BulkValue: Equatable {
+        /// All variations have the same value
+        case value(String)
+        /// When variations have mixed values.
+        case mixed
+        /// None of the variation has a value
+        case none
+    }
+
+    /// Calculates and returns if a property, specified by a `ProductVariation` keypath , has the same value or different values or is missing
+    /// over all product variations of this view model.
+    ///
+    func bulkValueOf(_ keypath: KeyPath<ProductVariation, String?>) -> BulkValue {
+        let allValues = productVariations.map(keypath)
+        let allValuesWithoutNilOrEmpty = allValues.compactMap { $0 }.filter { !$0.isEmpty }
+        // filter all unique values that are not nil or empty
+        let uniqueResults = allValuesWithoutNilOrEmpty.uniqued()
+
+        if let sameValue = uniqueResults.first {
+            if uniqueResults.count == 1 && allValues.count == allValuesWithoutNilOrEmpty.count {
+                // If we have only 1 value and we did not had any nil/empty
+                return BulkValue.value(sameValue)
+            } else {
+                // If at least ony value is different, even if it is missing (nil/empty)
+                return BulkValue.mixed
+            }
+        } else {
+            // If all values are missing or empty
+            return BulkValue.none
+        }
+    }
```

Comment:
Nice use of Keypaths! 

I think you can make this a bit more readable if you guard the first statement.

```swift
// Return `.none` when there are no values to evaluate.
guard let sameValue = uniqueResults.first else {
    return .none
}

if uniqueResults.count == 1 && allValues.count == allValuesWithoutNilOrEmpty.count {
    // If we have only 1 unique value and no empty/nil values, return `.value`
    return .value(sameValue)
} else {
    // return `.mixed` if at least only value is different, even if it is missing (nil/empty)
    return .mixed
}
```

I'm also curious about how you plan this type to scale when we need to compare more variation properties that aren't strings, for example, dimensions!

Reply:
> I think you can make this a bit more readable if you guard the first statement.

I agree, less nesting makes it more clear. I fixed it!

> I'm also curious about how you plan this type to scale when we need to compare more variation properties that aren't strings, for example, dimensions!

This would not be very hard. We need two things: 
* Replace the `String` type in the keyPath and `BulkValue` enum with a generic placeholder type(conforming to `Equatable` for finding same values).
* Properties like price and dimensions, when they are missing instead of nil they have empty strings as values. So the placeholder type would need to conform to a Protocol with a methdod like `isEmpty() -> Bool` (Strings already have this method).

So it would look something like this:

  ```Swift
  enum BulkValue<T: Equatable>: Equatable {
        /// All variations have the same value
        case value(T)
        /// When variations have mixed values.
        case mixed
        /// None of the variation has a value
        case none
    }

    func bulkValueOf<T>(_ keypath: KeyPath<ProductVariation, T>) -> BulkValue<T> 
     where T: Hashable, T:ValueContainer {
        let allValues = productVariations.map(keypath)
        let allValuesWithoutNilOrEmpty = allValues.compactMap { $0 }.filter { !$0.isEmpty() }
        // filter all unique values that are not nil or empty
        let uniqueResults = allValuesWithoutNilOrEmpty.uniqued()

        // Return `.none` when there are no values to evaluate.
        guard let sameValue = uniqueResults.first else {
            return .none
        }

        if uniqueResults.count == 1 && allValues.count == allValuesWithoutNilOrEmpty.count {
            // If we have only 1 value and we did not had any nil/empty
            return .value(sameValue)
        } else {
            // If at least ony value is different, even if it is missing (nil/empty)
            return .mixed
        }
    }

    /// Can do better on the name :)
    protocol ValueContainer {
        func isEmpty() -> Bool
    }
```
The `ProductDimensions` would need to conform to this to indicate when an instance has no value.

Reply:
Cool!

Reply:
Cool!